sklearn/model_selection/_search.py

```diff
@@ -688,10 +689,20 @@ def evaluate_candidates(candidate_params):
         # best_score_ iff refit is one of the scorer names
         # In single metric evaluation, refit_metric is "score"
         if self.refit or not self.multimetric_:
-            self.best_index_ = results["rank_test_%s" % refit_metric].argmin()
+            # If callable, refit is expected to return the index of the best
+            # parameter set.
+            if callable(self.refit):
+                self.best_index_ = self.refit(results)
+                if not isinstance(self.best_index_, (int, np.integer)):
+                    raise TypeError('best_index_ returned is not an integer')
+                if self.best_index_ < 0 or self.best_index_ >= len(results):
```

Comment:
Pretty sure this is a bug: results is a dictionary of things, and each value is an array the size of the grid.

Reply:
opened #13413